File: gen_drawio.dart
import 'dart:io';

void main() {
  final darkColors = [
        {"name": "cardBackground", "hex": "#1A2235"},
        {"name": "cardBorder", "hex": "#2A3750"},
        {"name": "titleColor", "hex": "#FFFFFF"},
        {"name": "subtitleColor", "hex": "#8A94A6"},
        {"name": "labelColor", "hex": "#00BCD4"},
        {"name": "inputFill", "hex": "#222B40"},
        {"name": "inputBorder", "hex": "#2A3040"},
        {"name": "inputFocusedBorder", "hex": "#00BCD4"},
        {"name": "inputHintText", "hex": "#8A94A6"},
        {"name": "inputTextColor", "hex": "#FFFFFF"},
        {"name": "inputIconColor", "hex": "#8A94A6"},
        {"name": "dropdownBackground", "hex": "#1A2235"},
        {"name": "dropdownIcon", "hex": "#00BCD4"},
        {"name": "modeButtonSelectedBackground", "hex": "#00BCD4"},
        {"name": "primaryButtonBackground", "hex": "#00BCD4"},
        {"name": "primaryButtonText", "hex": "#0D1118"},
        {"name": "pageBackground", "hex": "#0D1118"},
        {"name": "surfaceColor", "hex": "#1A2235"},
        {"name": "textOnSurface", "hex": "#FFFFFF"},
        {"name": "textSecondary", "hex": "#8A94A6"},
        {"name": "loadingIndicatorColor", "hex": "#00BCD4"},
        {"name": "appBarBackground", "hex": "#1C2030"},
        {"name": "appBarForeground", "hex": "#FFFFFF"},
        {"name": "appBarAccent", "hex": "#00BCD4"},
        {"name": "iconBoxBackground", "hex": "#222B40"},
        {"name": "iconBoxIcon", "hex": "#00BCD4"},
        {"name": "menuBackground", "hex": "#0D1118"},
        {"name": "menuBorder", "hex": "#2A3750"},
        {"name": "menuSelectedBackground", "hex": "#1C2030"},
        {"name": "menuSelectedIcon", "hex": "#00BCD4"},
        {"name": "dateTimeGradientStart", "hex": "#1A2235"},
        {"name": "dateTimeGradientEnd", "hex": "#0D1118"},
        {"name": "cardSurface", "hex": "#1A2235"},
        {"name": "cardBorderColor", "hex": "#2A3750"},
        {"name": "dialogBackground", "hex": "#1A2235"},
        {"name": "dividerColor", "hex": "#2A3040"},
        {"name": "mapGrid", "hex": "#162032"}
  ];

  final lightColors = [
        {"name": "cardBackground", "hex": "#FFFFFF"},
        {"name": "cardBorder", "hex": "#00BCD4"},
        {"name": "titleColor", "hex": "#1A1A2E"},
        {"name": "subtitleColor", "hex": "#7A8290"},
        {"name": "labelColor", "hex": "#00BCD4"},
        {"name": "inputFill", "hex": "#F5F7FA"},
        {"name": "inputBorder", "hex": "#D0D4DA"},
        {"name": "inputFocusedBorder", "hex": "#00BCD4"},
        {"name": "inputHintText", "hex": "#7A8290"},
        {"name": "inputTextColor", "hex": "#1A1A2E"},
        {"name": "inputIconColor", "hex": "#7A8290"},
        {"name": "dropdownBackground", "hex": "#FFFFFF"},
        {"name": "dropdownIcon", "hex": "#00BCD4"},
        {"name": "modeButtonSelectedBackground", "hex": "#00BCD4"},
        {"name": "primaryButtonBackground", "hex": "#00BCD4"},
        {"name": "primaryButtonText", "hex": "#FFFFFF"},
        {"name": "pageBackground", "hex": "#E8ECF0"},
        {"name": "surfaceColor", "hex": "#FFFFFF"},
        {"name": "textOnSurface", "hex": "#1A1A2E"},
        {"name": "textSecondary", "hex": "#7A8290"},
        {"name": "loadingIndicatorColor", "hex": "#00BCD4"},
        {"name": "appBarBackground", "hex": "#FFFFFF"},
        {"name": "appBarForeground", "hex": "#1A1A2E"},
        {"name": "appBarAccent", "hex": "#00BCD4"},
        {"name": "iconBoxBackground", "hex": "#F5F7FA"},
        {"name": "iconBoxIcon", "hex": "#00BCD4"},
        {"name": "menuBackground", "hex": "#FFFFFF"},
        {"name": "menuBorder", "hex": "#D0D4DA"},
        {"name": "menuSelectedBackground", "hex": "#F5F7FA"},
        {"name": "menuSelectedIcon", "hex": "#00BCD4"},
        {"name": "dateTimeGradientStart", "hex": "#F5F7FA"},
        {"name": "dateTimeGradientEnd", "hex": "#FFFFFF"},
        {"name": "cardSurface", "hex": "#FFFFFF"},
        {"name": "cardBorderColor", "hex": "#00BCD4"},
        {"name": "dialogBackground", "hex": "#FFFFFF"},
        {"name": "dividerColor", "hex": "#D0D4DA"},
        {"name": "mapGrid", "hex": "#D5DCE4"}
  ];

  List<String> xml = [
    '<mxfile version="22.1.2" type="device">', 
    '<diagram id="diagram1" name="AppTheme">', 
    '<mxGraphModel dx="1200" dy="1200" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="1200" pageHeight="2300" math="0" shadow="0">',
    '<root>',
    '<mxCell id="0" />',
    '<mxCell id="1" parent="0" />'
  ];

  int darkH = darkColors.length * 60 + 60;
  int lightH = lightColors.length * 60 + 60;

  xml.add('<mxCell id="dark_theme" value="DARK THEME (SCADA)" style="swimlane;whiteSpace=wrap;html=1;fillColor=#1C2030;fontColor=#FFFFFF;startSize=40;strokeColor=#2A3750;" vertex="1" parent="1"><mxGeometry x="50" y="50" width="300" height="$darkH" as="geometry" /></mxCell>');
  xml.add('<mxCell id="light_theme" value="LIGHT THEME (SCADA)" style="swimlane;whiteSpace=wrap;html=1;fillColor=#F5F7FA;fontColor=#1A1A2E;startSize=40;strokeColor=#D0D4DA;" vertex="1" parent="1"><mxGeometry x="400" y="50" width="300" height="$lightH" as="geometry" /></mxCell>');

  int yOffset = 60;
  for (int i=0; i<darkColors.length; i++) {
    String name = darkColors[i]["name"]!;
    String hex = darkColors[i]["hex"]!;
    String tColor = ["#00BCD4", "#FFFFFF"].contains(hex) ? "#000000" : "#FFFFFF";
    xml.add('<mxCell id="dark_$i" value="$name&#10;$hex" style="rounded=1;whiteSpace=wrap;html=1;fillColor=$hex;fontColor=$tColor;strokeColor=#FFFFFF;" vertex="1" parent="dark_theme"><mxGeometry x="30" y="$yOffset" width="240" height="40" as="geometry" /></mxCell>');
    yOffset += 50;
  }

  yOffset = 60;
  for (int i=0; i<lightColors.length; i++) {
    String name = lightColors[i]["name"]!;
    String hex = lightColors[i]["hex"]!;
    String tColor = ["#1A1A2E", "#7A8290"].contains(hex) ? "#FFFFFF" : "#000000";
    xml.add('<mxCell id="light_$i" value="$name&#10;$hex" style="rounded=1;whiteSpace=wrap;html=1;fillColor=$hex;fontColor=$tColor;strokeColor=#000000;" vertex="1" parent="light_theme"><mxGeometry x="30" y="$yOffset" width="240" height="40" as="geometry" /></mxCell>');
    yOffset += 50;
  }

  xml.addAll(['</root>', '</mxGraphModel>', '</diagram>', '</mxfile>']);
  
  final dir = Directory('c:/apps/toho_EGS/docs');
  if (!dir.existsSync()) {
    dir.createSync(recursive: true);
  }
  final file = File('c:/apps/toho_EGS/docs/app_theme.drawio');
  file.writeAsStringSync(xml.join('\n'));
  print('Done');
}

File: gen_drawio.py
import os

def gen_drawio():
    dark_colors = [
        ("cardBackground", "#1E241E"),
        ("cardBorder", "#1AFFFFFF"),
        ("titleColor", "#FFFFFF"),
        ("subtitleColor", "#8AFFFFFF"),
        ("labelColor", "#69F0AE"),
        ("inputFill", "#0DFFFFFF"),
        ("inputBorder", "#3DFFFFFF"),
        ("inputFocusedBorder", "#00FF00"),
        ("inputHintText", "#3DFFFFFF"),
        ("inputTextColor", "#FFFFFF"),
        ("inputIconColor", "#8AFFFFFF"),
        ("dropdownBackground", "#1E241E"),
        ("modeButtonSelectedBackground", "#00C853"),
        ("primaryButtonBackground", "#00C853"),
        ("primaryButtonText", "#000000"),
        ("pageBackground", "#0F1410"),
        ("surfaceColor", "#1E293B"),
        ("textOnSurface", "#FFFFFF"),
        ("textSecondary", "#B0BEC5"),
        ("loadingIndicatorColor", "#2ECC71"),
        ("appBarBackground", "#0F1410"),
        ("appBarForeground", "#FFFFFF"),
        ("appBarAccent", "#2ECC71"),
        ("iconBoxBackground", "#1E3A2A"),
        ("iconBoxIcon", "#2ECC71"),
        ("menuBackground", "#0F1410"),
        ("menuBorder", "#1E3A2A"),
        ("menuSelectedBackground", "#1E3A2A"),
        ("menuSelectedIcon", "#2ECC71"),
        ("dateTimeGradientStart", "#1E3A2A"),
        ("dateTimeGradientEnd", "#0F1410"),
        ("cardSurface", "#1E293B"),
        ("cardBorderColor", "#1E3A2A"),
        ("dialogBackground", "#1E293B"),
        ("dividerColor", "#1E3A2A"),
    ]

    light_colors = [
        ("cardBackground", "#FFFBF5"),
        ("cardBorder", "#FFD9A0"),
        ("titleColor", "#3E2723"),
        ("subtitleColor", "#795548"),
        ("labelColor", "#E65100"),
        ("inputFill", "#FFF3E0"),
        ("inputBorder", "#FFCC80"),
        ("inputFocusedBorder", "#F57C00"),
        ("inputHintText", "#BCAAA4"),
        ("inputTextColor", "#3E2723"),
        ("inputIconColor", "#795548"),
        ("dropdownBackground", "#FFF8EF"),
        ("modeButtonSelectedBackground", "#F57C00"),
        ("primaryButtonBackground", "#F57C00"),
        ("primaryButtonText", "#FFFFFF"),
        ("pageBackground", "#FFF8F0"),
        ("surfaceColor", "#FFFFFF"),
        ("textOnSurface", "#3E2723"),
        ("textSecondary", "#8D6E63"),
        ("loadingIndicatorColor", "#F57C00"),
        ("appBarBackground", "#FFF8F0"),
        ("appBarForeground", "#3E2723"),
        ("appBarAccent", "#F57C00"),
        ("iconBoxBackground", "#FFE0B2"),
        ("iconBoxIcon", "#F57C00"),
        ("menuBackground", "#FFF3E0"),
        ("menuBorder", "#FFD9A0"),
        ("menuSelectedBackground", "#FFE0B2"),
        ("menuSelectedIcon", "#F57C00"),
        ("dateTimeGradientStart", "#FFE0B2"),
        ("dateTimeGradientEnd", "#FFF3E0"),
        ("cardSurface", "#FFFFFF"),
        ("cardBorderColor", "#FFCC80"),
        ("dialogBackground", "#FFFBF5"),
        ("dividerColor", "#FFCC80"),
    ]

    xml = ['<mxfile version="22.1.2" type="device">', '<diagram id="diagram1" name="AppTheme">', 
           '<mxGraphModel dx="1200" dy="1200" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="1200" pageHeight="1600" math="0" shadow="0">',
           '<root>',
           '<mxCell id="0" />',
           '<mxCell id="1" parent="0" />']

    # Swimlanes
    xml.append(f'<mxCell id="dark_theme" value="DARK THEME (GREEN)" style="swimlane;whiteSpace=wrap;html=1;fillColor=#1E241E;fontColor=#FFFFFF;startSize=40;strokeColor=#1AFFFFFF;" vertex="1" parent="1"><mxGeometry x="50" y="50" width="300" height="{len(dark_colors)*60+60}" as="geometry" /></mxCell>')
    xml.append(f'<mxCell id="light_theme" value="LIGHT THEME (ORANGE)" style="swimlane;whiteSpace=wrap;html=1;fillColor=#FFFBF5;fontColor=#3E2723;startSize=40;strokeColor=#FFD9A0;" vertex="1" parent="1"><mxGeometry x="400" y="50" width="300" height="{len(light_colors)*60+60}" as="geometry" /></mxCell>')

    # Add dark colors
    y_offset = 60
    for i, (name, hex_c) in enumerate(dark_colors):
        # determine text color based on hex
        t_color = "#FFFFFF" if not hex_c in ["#00FF00", "#00C853", "#FFFFFF", "#2ECC71"] else "#000000"
        node_id = f"dark_{i}"
        xml.append(f'<mxCell id="{node_id}" value="{name}&#10;{hex_c}" style="rounded=1;whiteSpace=wrap;html=1;fillColor={hex_c};fontColor={t_color};strokeColor=#FFFFFF;" vertex="1" parent="dark_theme"><mxGeometry x="30" y="{y_offset}" width="240" height="40" as="geometry" /></mxCell>')
        y_offset += 50

    # Add light colors
    y_offset = 60
    for i, (name, hex_c) in enumerate(light_colors):
        t_color = "#000000" if hex_c not in ["#3E2723", "#F57C00", "#795548", "#8D6E63", "#E65100"] else "#FFFFFF"
        node_id = f"light_{i}"
        xml.append(f'<mxCell id="{node_id}" value="{name}&#10;{hex_c}" style="rounded=1;whiteSpace=wrap;html=1;fillColor={hex_c};fontColor={t_color};strokeColor=#000000;" vertex="1" parent="light_theme"><mxGeometry x="30" y="{y_offset}" width="240" height="40" as="geometry" /></mxCell>')
        y_offset += 50

    xml.extend(['</root>', '</mxGraphModel>', '</diagram>', '</mxfile>'])
    
    out_dir = r"c:\apps\toho_EGS\docs"
    if not os.path.exists(out_dir):
        os.makedirs(out_dir)
        
    with open(os.path.join(out_dir, "app_theme.drawio"), "w") as f:
        f.write("\\n".join(xml))

if __name__ == "__main__":
    gen_drawio()
    print("Done generating c:\\apps\\toho_EGS\\docs\\app_theme.drawio")
